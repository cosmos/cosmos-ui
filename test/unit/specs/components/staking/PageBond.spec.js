import setup from "../../../helpers/vuex-setup"
import htmlBeautify from "html-beautify"
import Vuelidate from "vuelidate"
import PageBond from "renderer/components/staking/PageBond"
import interact from "interactjs"

describe("PageBond", () => {
  let wrapper, store, router
  let { mount, localVue } = setup()
  localVue.use(Vuelidate)

  beforeEach(() => {
    let test = mount(PageBond, {
      doBefore: ({ store }) => {
        store.commit("setAtoms", 101)

        store.commit("addToCart", {
          id: "pubkeyX",
          pub_key: {
            type: "ed25519",
            data: "pubkeyX"
          },
          voting_power: 10000,
          shares: 5000,
          description: "descriptionX",
          country: "USA",
          moniker: "someValidator"
        })
        store.commit("addToCart", {
          id: "pubkeyY",
          pub_key: {
            type: "ed25519",
            data: "pubkeyY"
          },
          voting_power: 30000,
          shares: 10000,
          description: "descriptionY",
          country: "Canada",
          moniker: "someOtherValidator"
        })
      }
    })
    store = test.store
    router = test.router
    wrapper = test.wrapper

    wrapper.update()
  })

  it("has the expected html structure", async () => {
    // after importing the @tendermint/ui components from modules
    // the perfect scroll plugin needs a $nextTick and a wrapper.update
    // to work properly in the tests (snapshots weren't matching)
    // this has occured across multiple tests
    await wrapper.vm.$nextTick()
    wrapper.update()
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it("shows number of total atoms", () => {
    store.commit("setAtoms", 1337)
    expect(wrapper.vm.totalAtoms).toBe(1337)
  })

  it("shows old bonded atoms ", () => {
    store.commit("setCommittedDelegation", {
      candidateId: "pubkeyX",
      value: 13
    })
    store.commit("setCommittedDelegation", {
      candidateId: "pubkeyY",
      value: 26
    })
    expect(wrapper.vm.oldBondedAtoms).toBe(39)
  })

  it("shows bond bar percent", () => {
    store.commit("setAtoms", 120)
    expect(wrapper.vm.bondBarPercent(30)).toBe("25%")
  })

  it("sets bond bar inner width and style", () => {
    store.commit("setAtoms", 120)
    wrapper.setData({ bondBarOuterWidth: 128 })
    expect(wrapper.vm.bondBarInnerWidth(80)).toBe("95px")
    expect(wrapper.vm.styleBondBarInner(80)).toEqual({ width: "95px" })
  })

  it("sets bond group class", () => {
    expect(wrapper.vm.bondGroupClass(1337)).toBe("bond-group--positive")
    expect(wrapper.vm.bondGroupClass(-1337)).toBe("bond-group--negative")
    expect(wrapper.vm.bondGroupClass(0)).toBe("bond-group--neutral")
  })

  it("has bond bars the user can interact with", () => {
    expect(interact.isSet(".bond-bar__inner--editable")).toBe(true)
  })

  it("updates delegate atoms", () => {
    wrapper.vm.updateDelegateAtoms("pubkeyX", 88)
    let delegate = wrapper.vm.fields.delegates.find(d => d.id === "pubkeyX")
    expect(delegate.atoms).toBe(88)
  })

  it("only shows percent based on showing atoms", () => {
    let updatedDelegate = wrapper.vm.handleResize(
      wrapper.vm.$el.querySelector("#delegate-pubkeyX"),
      0.12
    )
    expect(updatedDelegate.deltaAtomsPercent).toBe("100%")
  })

  it("calculates delta", () => {
    expect(wrapper.vm.delta(100.23293423, 90.5304934)).toBe(9.70244083)
    expect(wrapper.vm.delta(100, 90, "int")).toBe(10)
  })

  it("calculates percentages", () => {
    expect(wrapper.vm.percent(45, 60)).toBe("75%")
    expect(wrapper.vm.percent(40, 60, 4)).toBe("66.6667%")
  })

  it("limits the input of atoms to the maximum", () => {
    let delegate = {
      id: "pubkeyX",
      delegate: store.getters.shoppingCart[0].delegate,
      atoms: 50,
      oldAtoms: 40
    }
    wrapper.vm.limitMax(delegate, 100)
    expect(delegate.atoms).toBe(50)

    wrapper.vm.limitMax(delegate, 10)
    expect(delegate.atoms).toBe(10)
  })

  it("leaves if there are no candidates selected", () => {
    let { router } = mount(PageBond, {
      doBefore: ({ store }) => {
        store.commit("setAtoms", 101)
      }
    })
    expect(router.currentRoute.fullPath).toBe("/staking")
  })

  it("leaves if no atoms available", () => {
    let test = mount(PageBond, {
      doBefore: ({ store }) => {
        store.commit("setAtoms", 0)
      }
    })
    router = test.router
    expect(router.currentRoute.fullPath).toBe("/staking")
  })

  it("shows selected candidates", () => {
    expect(htmlBeautify(wrapper.html())).toContain("someValidator")
    expect(htmlBeautify(wrapper.html())).toContain("someOtherValidator")
  })

  it("resets fields properly", () => {
    wrapper.setData({
      fields: {
        delegates: [
          {
            id: "pubkeyX",
            delegate: store.getters.shoppingCart[0].delegate,
            atoms: 50,
            oldAtoms: 40
          },
          {
            id: "pubkeyY",
            delegate: store.getters.shoppingCart[1].delegate,
            atoms: 50,
            oldAtoms: 40
          }
        ]
      }
    })
    expect(wrapper.find("#new-unbonded-atoms").element.value).toBe("81")
    wrapper.find("#btn-reset").trigger("click")
    expect(wrapper.find("#new-unbonded-atoms").element.value).toBe("101")
  })

  it("shows an error when bonding too many atoms", () => {
    wrapper.setData({
      fields: {
        delegates: [
          {
            id: "pubkeyX",
            delegate: store.getters.shoppingCart[0].delegate,
            atoms: 100
          },
          {
            id: "pubkeyY",
            delegate: store.getters.shoppingCart[1].delegate,
            atoms: 100
          }
        ]
      }
    })
    wrapper.findAll("#btn-bond").trigger("click")
    expect(store.dispatch.mock.calls[0]).toBeUndefined()
    expect(wrapper.vm.$el.querySelector(".tm-form-msg--error")).not.toBeNull()
  })

  it("shows an appropriate amount of unbonded atoms", () => {
    wrapper.setData({
      fields: {
        bondConfirm: false,
        delegates: [
          {
            id: "pubkeyX",
            delegate: store.getters.shoppingCart[0].delegate,
            atoms: 30,
            oldAtoms: 20
          },
          {
            id: "pubkeyY",
            delegate: store.getters.shoppingCart[1].delegate,
            atoms: 30,
            oldAtoms: 20
          }
        ]
      }
    })
    expect(wrapper.vm.newUnbondedAtoms).toBe(81)
    expect(wrapper.find("#new-unbonded-atoms").vnode.elm._value).toBe(81)
  })

  it("shows an appropriate amount of unbonding atoms", () => {
    // give some stake so unbinding bar shows
    store.commit("setCommittedDelegation", { candidateId: "pubkeyX", value: 5 })
    wrapper.update()

    wrapper.setData({
      fields: {
        bondConfirm: false,
        delegates: [
          {
            id: "pubkeyX",
            delegate: store.getters.shoppingCart[0].delegate,
            atoms: 31,
            oldAtoms: 41
          },
          {
            id: "pubkeyY",
            delegate: store.getters.shoppingCart[1].delegate,
            atoms: 30,
            oldAtoms: 40
          }
        ]
      }
    })
    expect(wrapper.vm.newUnbondingAtoms).toBe(20)
    expect(wrapper.find("#new-unbonding-atoms").vnode.elm._value).toBe(20)
  })

  it("shows an error if confirmation is not checked", () => {
    wrapper.setData({
      fields: {
        bondConfirm: false,
        delegates: [
          {
            id: "pubkeyX",
            delegate: store.getters.shoppingCart[0].delegate,
            atoms: 51
          },
          {
            id: "pubkeyY",
            delegate: store.getters.shoppingCart[1].delegate,
            atoms: 50
          }
        ]
      }
    })
    wrapper.findAll("#btn-bond").trigger("click")
    expect(store.dispatch.mock.calls[0]).toBeUndefined()
    expect(wrapper.vm.$el.querySelector(".tm-form-msg--error")).not.toBeNull()
  })

  it("submits a bonding action on submit", async () => {
    wrapper.setData({
      fields: {
        bondConfirm: true,
        delegates: [
          {
            id: "pubkeyX",
            delegate: store.getters.shoppingCart[0].delegate,
            atoms: 51
          },
          {
            id: "pubkeyY",
            delegate: store.getters.shoppingCart[1].delegate,
            atoms: 50
          }
        ]
      }
    })
    wrapper.findAll("#btn-bond").trigger("click")
    expect(store.getDispatches("submitDelegation")).toHaveLength(1)
    // XXX somehow this still shows the error in tests but this does not happen in live Voyager
    // XXX this makes the .not.toBeNull() tests pointless :/
    // expect(wrapper.vm.$el.querySelector(".tm-form-msg--error")).toBeNull()
  })

  it("moves to the staking page after bonding", async () => {
    store._actions.sendTx[0] = () => Promise.resolve()
    wrapper.setData({
      fields: {
        bondConfirm: true,
        delegates: [
          {
            id: "pubkeyX",
            delegate: store.getters.shoppingCart[0].delegate,
            atoms: 51
          },
          {
            id: "pubkeyY",
            delegate: store.getters.shoppingCart[1].delegate,
            atoms: 50
          }
        ]
      }
    })
    await wrapper.vm.onSubmit()
    expect(store.getCommits("notify")).toHaveLength(1)
    expect(router.currentRoute.fullPath).toBe("/staking")
  })

  it("shows an error if unbonding too many atoms", async () => {
    wrapper.setData({
      fields: {
        bondConfirm: false,
        delegates: [
          {
            id: "pubkeyX",
            delegate: store.getters.shoppingCart[0].delegate,
            atoms: 51,
            oldAtoms: 41
          },
          {
            id: "pubkeyY",
            delegate: store.getters.shoppingCart[1].delegate,
            atoms: 241,
            oldAtoms: 40
          }
        ]
      }
    })
    expect(wrapper.vm.newUnbondedAtoms).toBeLessThan(0)
    await wrapper.vm.onSubmit()
    expect(store.dispatch.mock.calls[0]).toBeUndefined()
    expect(
      store.commit.mock.calls.filter(
        x =>
          x[0] === "notifyError" &&
          x[1].body.indexOf("more atoms than you have") !== -1
      )
    ).toBeDefined()
    expect(wrapper.vm.$el.querySelector(".tm-form-msg--error")).not.toBeNull()
  })

  it("unbonds atoms if bond amount is decreased", () => {
    store.commit("setCommittedDelegation", {
      candidateId: "pubkeyX",
      value: 51
    })
    store.commit("setCommittedDelegation", {
      candidateId: "pubkeyY",
      value: 50
    })
    wrapper.update()
    wrapper.setData({
      fields: {
        bondConfirm: true,
        delegates: [
          {
            id: "pubkeyX",
            delegate: store.getters.shoppingCart[0].delegate,
            atoms: 0
          },
          {
            id: "pubkeyY",
            delegate: store.getters.shoppingCart[1].delegate,
            atoms: 25
          }
        ]
      }
    })
    wrapper.findAll("#btn-bond").trigger("click")
    expect(store.dispatch.mock.calls[0][0]).toBe("submitDelegation")
  })

  it("does not show the unbonding bar if user has no atoms", () => {
    expect(wrapper.vm.oldBondedAtoms).toBe(0)
    expect(
      wrapper.vm.$el.querySelector(".bond-group.bond-group--unbonding")
    ).toBeNull()
  })

  it("shows a message if there are revoked candidates", () => {
    wrapper.setData({
      fields: {
        delegates: [
          {
            id: "pubkeyX",
            delegate: Object.assign(
              {},
              store.getters.shoppingCart[0].delegate,
              { revoked: true }
            ),
            atoms: 0
          },
          {
            id: "pubkeyY",
            delegate: store.getters.shoppingCart[1].delegate,
            atoms: 25
          }
        ]
      }
    })
    expect(wrapper.vm.showsRevokedValidators).toBe(true)
    expect(wrapper.vm.$el).toMatchSnapshot()
  })
})
