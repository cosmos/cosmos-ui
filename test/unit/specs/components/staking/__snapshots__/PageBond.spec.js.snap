// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageBond has the expected html structure 1`] = `
<div
  class="tm-page page-bond"
>
  <header
    class="tm-page-header"
  >
    <div
      class="tm-page-header-container"
    >
      <div
        class="tm-page-header-text"
      >
        <div
          class="tm-page-header-title"
        >
          <h2>
            Bond STEAK
          </h2>
        </div>
        <div
          class="tm-page-header-subtitle"
        >
          <h3>
            
          </h3>
        </div>
      </div>
      <menu
        class="tm-page-header-menu"
      >
        <div>
          <div>
            <div
              class="tm-tool-bar"
            >
              <div
                class="tm-tool-bar-container"
              >
                <div
                  class="main"
                >
                  <a
                    class="back"
                    disabled="disabled"
                  >
                    <i
                      class="material-icons"
                    >
                      arrow_back
                    </i>
                  </a>
                  <a
                    class="help"
                  >
                    <i
                      class="material-icons"
                    >
                      help_outline
                    </i>
                  </a>
                  <a
                    class="sign-out"
                  >
                    <i
                      class="material-icons"
                    >
                      exit_to_app
                    </i>
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </menu>
    </div>
  </header>
  <main
    class="tm-page-main ps"
  >
    <section
      class="tm-part"
    >
      <div
        class="tm-part-container"
      >
        <header
          class="tm-part-header"
        >
          <div
            class="tm-part-header-container"
          >
            <div
              class="tm-part-title h5"
            >
              Start bonding your 101 STEAK
            </div>
            <menu
              class="tm-part-menu"
            />
          </div>
        </header>
        <main
          class="tm-part-main"
        >
          <form
            class="tm-form"
          >
            <div
              class="tm-form-container"
            >
              <!---->
              <main
                class="tm-form-main"
              >
                <div
                  class="bond-group bond-group--neutral"
                >
                  <div
                    class="bond-group__fields"
                  >
                    <div
                      class="bond-bar"
                    >
                      <label
                        class="bond-bar__label"
                      >
                        Unbonded STEAK
                      </label>
                      <div
                        class="bond-bar__input"
                      >
                        <div
                          class="bond-bar-old__outer"
                        >
                          <div
                            class="bond-bar-old__inner"
                            style="width: 0px;"
                          />
                        </div>
                        <div
                          class="bond-bar__outer"
                        >
                          <div
                            class="bond-bar__inner"
                            style="width: 0px;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="bond-percent"
                    >
                      <label
                        class="bond-delta"
                      >
                        <!---->
                      </label>
                      <input
                        class="bond-percent__input tm-field"
                        disabled="disabled"
                        placeholder="0%"
                      />
                    </div>
                    <div
                      class="bond-value"
                    >
                      <label
                        class="bond-delta"
                      >
                        <!---->
                      </label>
                      <input
                        class="bond-value__input tm-field"
                        disabled="disabled"
                        id="new-unbonded-atoms"
                        placeholder="Atoms"
                        type="number"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="bond-group bond-candidate bond-group--neutral"
                >
                  <div
                    class="bond-group__fields"
                  >
                    <div
                      class="bond-bar"
                    >
                      <label
                        class="bond-bar__label"
                      >
                        someValidator
                      </label>
                      <!---->
                      <!---->
                      <div
                        class="bond-bar__input"
                      >
                        <div
                          class="bond-bar-old__outer"
                        >
                          <!---->
                        </div>
                        <div
                          class="bond-bar__outer"
                        >
                          <div
                            class="bond-bar__inner bond-bar__inner--editable"
                            id="delegate-pubkeyX"
                            style="width: 28px;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="bond-percent"
                    >
                      <label
                        class="bond-delta"
                      >
                        <!---->
                      </label>
                      <input
                        class="bond-percent__input tm-field"
                        disabled="disabled"
                        placeholder="0%"
                      />
                    </div>
                    <div
                      class="bond-value"
                    >
                      <label
                        class="bond-delta"
                      >
                        <!---->
                      </label>
                      <input
                        class="bond-value__input tm-field"
                        min="0"
                        placeholder="Atoms"
                        step="1"
                        type="number"
                      />
                    </div>
                  </div>
                  <!---->
                  <!---->
                  <!---->
                </div>
                <div
                  class="bond-group bond-candidate bond-group--neutral"
                >
                  <div
                    class="bond-group__fields"
                  >
                    <div
                      class="bond-bar"
                    >
                      <label
                        class="bond-bar__label"
                      >
                        someOtherValidator
                      </label>
                      <!---->
                      <!---->
                      <div
                        class="bond-bar__input"
                      >
                        <div
                          class="bond-bar-old__outer"
                        >
                          <!---->
                        </div>
                        <div
                          class="bond-bar__outer"
                        >
                          <div
                            class="bond-bar__inner bond-bar__inner--editable"
                            id="delegate-pubkeyY"
                            style="width: 28px;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="bond-percent"
                    >
                      <label
                        class="bond-delta"
                      >
                        <!---->
                      </label>
                      <input
                        class="bond-percent__input tm-field"
                        disabled="disabled"
                        placeholder="0%"
                      />
                    </div>
                    <div
                      class="bond-value"
                    >
                      <label
                        class="bond-delta"
                      >
                        <!---->
                      </label>
                      <input
                        class="bond-value__input tm-field"
                        min="0"
                        placeholder="Atoms"
                        step="1"
                        type="number"
                      />
                    </div>
                  </div>
                  <!---->
                  <!---->
                  <!---->
                </div>
                <!---->
                <div
                  class="tm-form-group"
                >
                  <!---->
                  <!---->
                  <div
                    class="tm-form-group__field"
                  >
                    <!---->
                    <div
                      class="tm-field-checkbox"
                    >
                      <div
                        class="tm-field-checkbox-input"
                      >
                        <input
                          id="bond-confirm"
                          type="checkbox"
                        />
                      </div>
                      <label
                        class="tm-field-checkbox-label"
                        for="bond-confirm"
                      >
                        Yes, update my bonds. I understand unbonding will take 30 days.
                      </label>
                    </div>
                    <div
                      class="tm-form-msg sm tm-form-msg--error"
                    >
                      Bonding Confirmation is required
                    </div>
                  </div>
                </div>
              </main>
              <footer
                class="tm-form-footer"
              >
                <div>
                  <button
                    class="tm-btn"
                    id="btn-reset"
                    type="button"
                  >
                    <span
                      class="tm-btn__container tm-btn--danger"
                    >
                      <!---->
                      <!---->
                      <span
                        class="tm-btn__value"
                      >
                        Reset
                      </span>
                    </span>
                  </button>
                  <button
                    class="tm-btn"
                    id="btn-bond"
                  >
                    <span
                      class="tm-btn__container tm-btn--primary"
                    >
                      <!---->
                      <!---->
                      <span
                        class="tm-btn__value"
                      >
                        Submit
                      </span>
                    </span>
                  </button>
                </div>
              </footer>
            </div>
          </form>
        </main>
      </div>
    </section>
    <div
      class="ps__rail-x"
      style="left: 0px; top: 0px;"
    >
      <div
        class="ps__thumb-x"
        style="left: 0px; width: 0px;"
        tabindex="0"
      />
    </div>
    <div
      class="ps__rail-y"
      style="top: 0px; left: 0px;"
    >
      <div
        class="ps__thumb-y"
        style="top: 0px; height: 0px;"
        tabindex="0"
      />
    </div>
  </main>
</div>
`;

exports[`PageBond shows a message if there are revoked candidates 1`] = `
<div
  class="tm-page page-bond"
>
  <header
    class="tm-page-header"
  >
    <div
      class="tm-page-header-container"
    >
      <div
        class="tm-page-header-text"
      >
        <div
          class="tm-page-header-title"
        >
          <h2>
            Bond STEAK
          </h2>
        </div>
        <div
          class="tm-page-header-subtitle"
        >
          <h3>
            
          </h3>
        </div>
      </div>
      <menu
        class="tm-page-header-menu"
      >
        <div>
          <div>
            <div
              class="tm-tool-bar"
            >
              <div
                class="tm-tool-bar-container"
              >
                <div
                  class="main"
                >
                  <a
                    class="back"
                  >
                    <i
                      class="material-icons"
                    >
                      arrow_back
                    </i>
                  </a>
                  <a
                    class="help"
                  >
                    <i
                      class="material-icons"
                    >
                      help_outline
                    </i>
                  </a>
                  <a
                    class="sign-out"
                  >
                    <i
                      class="material-icons"
                    >
                      exit_to_app
                    </i>
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </menu>
    </div>
  </header>
  <main
    class="tm-page-main"
  >
    <section
      class="tm-part"
    >
      <div
        class="tm-part-container"
      >
        <header
          class="tm-part-header"
        >
          <div
            class="tm-part-header-container"
          >
            <div
              class="tm-part-title h5"
            >
              Start bonding your 101 STEAK
            </div>
            <menu
              class="tm-part-menu"
            />
          </div>
        </header>
        <main
          class="tm-part-main"
        >
          <form
            class="tm-form"
          >
            <div
              class="tm-form-container"
            >
              <!---->
              <main
                class="tm-form-main"
              >
                <div
                  class="bond-group bond-group--neutral"
                >
                  <div
                    class="bond-group__fields"
                  >
                    <div
                      class="bond-bar"
                    >
                      <label
                        class="bond-bar__label"
                      >
                        Unbonded STEAK
                      </label>
                      <div
                        class="bond-bar__input"
                      >
                        <div
                          class="bond-bar-old__outer"
                        >
                          <div
                            class="bond-bar-old__inner"
                            style="width: 0px;"
                          />
                        </div>
                        <div
                          class="bond-bar__outer"
                        >
                          <div
                            class="bond-bar__inner"
                            style="width: 0px;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="bond-percent"
                    >
                      <label
                        class="bond-delta"
                      >
                        <!---->
                      </label>
                      <input
                        class="bond-percent__input tm-field"
                        disabled="disabled"
                        placeholder="0%"
                      />
                    </div>
                    <div
                      class="bond-value"
                    >
                      <label
                        class="bond-delta"
                      >
                        <!---->
                      </label>
                      <input
                        class="bond-value__input tm-field"
                        disabled="disabled"
                        id="new-unbonded-atoms"
                        placeholder="Atoms"
                        type="number"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="bond-group bond-candidate bond-group--neutral"
                >
                  <div
                    class="bond-group__fields"
                  >
                    <div
                      class="bond-bar"
                    >
                      <!---->
                      <label
                        class="bond-bar__label revoked"
                      >
                        someValidator
                      </label>
                      <label
                        class="bond-bar__revoked"
                      >
                        REVOKED
                      </label>
                      <div
                        class="bond-bar__input"
                      >
                        <div
                          class="bond-bar-old__outer"
                        >
                          <!---->
                        </div>
                        <div
                          class="bond-bar__outer"
                        >
                          <div
                            class="bond-bar__inner bond-bar__inner--editable"
                            id="delegate-pubkeyX"
                            style="width: 28px;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="bond-percent"
                    >
                      <label
                        class="bond-delta"
                      >
                        <span>
                          
                        </span>
                      </label>
                      <input
                        class="bond-percent__input tm-field"
                        disabled="disabled"
                        placeholder="0%"
                      />
                    </div>
                    <div
                      class="bond-value"
                    >
                      <label
                        class="bond-delta"
                      >
                        <span>
                          
                        </span>
                      </label>
                      <input
                        class="bond-value__input tm-field"
                        min="0"
                        placeholder="Atoms"
                        step="1"
                        type="number"
                      />
                    </div>
                  </div>
                  <!---->
                  <!---->
                  <!---->
                </div>
                <div
                  class="bond-group bond-candidate bond-group--neutral"
                >
                  <div
                    class="bond-group__fields"
                  >
                    <div
                      class="bond-bar"
                    >
                      <label
                        class="bond-bar__label"
                      >
                        someOtherValidator
                      </label>
                      <!---->
                      <!---->
                      <div
                        class="bond-bar__input"
                      >
                        <div
                          class="bond-bar-old__outer"
                        >
                          <!---->
                        </div>
                        <div
                          class="bond-bar__outer"
                        >
                          <div
                            class="bond-bar__inner bond-bar__inner--editable"
                            id="delegate-pubkeyY"
                            style="width: 21px;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="bond-percent"
                    >
                      <label
                        class="bond-delta"
                      >
                        <span>
                          
                        </span>
                      </label>
                      <input
                        class="bond-percent__input tm-field"
                        disabled="disabled"
                        placeholder="0%"
                      />
                    </div>
                    <div
                      class="bond-value"
                    >
                      <label
                        class="bond-delta"
                      >
                        <span>
                          
                        </span>
                      </label>
                      <input
                        class="bond-value__input tm-field"
                        min="0"
                        placeholder="Atoms"
                        step="1"
                        type="number"
                      />
                    </div>
                  </div>
                  <!---->
                  <!---->
                  <!---->
                </div>
                <!---->
                <div
                  class="tm-form-group"
                >
                  <!---->
                  <!---->
                  <div
                    class="tm-form-group__field"
                  >
                    <div
                      class="tm-form-msg sm tm-form-msg--desc"
                    >
                      A revoked validator is not validating and therefor is not producing rewards. The revoked state may be temporary.
                    </div>
                    <div
                      class="tm-field-checkbox"
                    >
                      <div
                        class="tm-field-checkbox-input"
                      >
                        <input
                          id="bond-confirm"
                          type="checkbox"
                        />
                      </div>
                      <label
                        class="tm-field-checkbox-label"
                        for="bond-confirm"
                      >
                        Yes, update my bonds. I understand unbonding will take 30 days.
                      </label>
                    </div>
                    <div
                      class="tm-form-msg sm tm-form-msg--error"
                    >
                      Bonding Confirmation is required
                    </div>
                  </div>
                </div>
              </main>
              <footer
                class="tm-form-footer"
              >
                <div>
                  <button
                    class="tm-btn"
                    id="btn-reset"
                    type="button"
                  >
                    <span
                      class="tm-btn__container tm-btn--danger"
                    >
                      <!---->
                      <!---->
                      <span
                        class="tm-btn__value"
                      >
                        Reset
                      </span>
                    </span>
                  </button>
                  <button
                    class="tm-btn"
                    id="btn-bond"
                  >
                    <span
                      class="tm-btn__container tm-btn--primary"
                    >
                      <!---->
                      <!---->
                      <span
                        class="tm-btn__value"
                      >
                        Submit
                      </span>
                    </span>
                  </button>
                </div>
              </footer>
            </div>
          </form>
        </main>
      </div>
    </section>
  </main>
</div>
`;
